### DISCUSSION TITLE
add-track no upload option

### DISCUSSION THREAD

**Discussion by cxbiosyn** [GMOD/jbrowse-components#4453] [Q&A] [answered]
Dear all, 

I am very new to Jbrowse2, learning to use by this tutorial 
http://gmod.org/wiki/JBrowse_2_Tutorial_PAG_2022#Adding_gene_annotations

When I try to add a GFF track, I see there are 4 options: URL,upload local file, and 2 cloud option 
![image](https://github.com/GMOD/jbrowse-components/assets/61400359/fadedba3-e18b-46c2-b1d9-497c81ad1740)

but in my own case, there is only the URL option
<img width="361" alt="image" src="https://github.com/GMOD/jbrowse-components/assets/61400359/89cd9f3d-4ec7-4fb6-a3e8-f8350bb56ab0">

I am using the web jbrowse2, I installed it in my server, I am not sure if I did everything correct

Regards
Chenxi

**Comment by cmdcolin** [comment, marked as answer]
Hi there :) you have done everything correct most likely

There are two factors at play:

1) If you are using the admin-server, it intentionally hides the "File" option. The admin-server tries to persist all changes to the config.json file, and we haven't properly implemented a "local file uploader" feature for admin-server. The File normally just opens the file locally in the browser, but admin-server wants to persist all the changes to the server, which is only possible with URL right now

2) Setting up the cloud options (dropbox, google drive) are actually a bit advanced for web instances, because these services require you to register a bunch of info with them to get their OAuth flow to work right. but to skip all the details, it's basically normal that you don't see these options, but if you did want them  let us know. it's a bit complicated to set it up but we could try to advise

**Reply by cmdcolin** [reply]
just as a final addendum: i would use the "jbrowse CLI" to actually add your files to the config for most cases currently

**Comment by cxbiosyn** [comment]
thank you for the quick reply!
I actually didn't want to use the "upload file" option, but I had a trouble using URL currently

I have a website based on wordpress(https://xliugroup.com/test/), and I upload 3 files(.gz/.fai/.gzi) from Klebsiella pneumoniae NTUH-K2044, I download the genome from NCBI, its .fna, I just changed the suffix .fna to .fa, 

1. I downloaded Klebsiella pneumoniae NTUH-K2044 genome file from NCBI (.fa and .gff)

2. I changed the suffix .fna to .fa, and used Samtools to get .gz, .gzi, and .fai

3. I uploaded these 3 files in our lab website(based on wordpress), if I click the URL, I can download them (https://xliugroup.com/test/)

4. I followed jbrowse2 GUI protocol((https://jbrowse.org/jb2/docs/tutorials/config_gui/)) to config them by their URL, and I got error, probably something wrong with the URL I used?
<img width="1429" alt="image" src="https://github.com/GMOD/jbrowse-components/assets/61400359/8e229da0-6ca8-4706-a386-4f0f5a1495ad">

5. if I use the URL from the hg38 example, I can get the correct assembly
6. then I tried to use "File" option for assembly NTUH-K2044, it worked

that's why I am asking the gff "file oploader" option
I will learn to use the jbrowse-cli, but the URL problem still there...

regards
Chenxi

**Reply by cmdcolin** [reply]
the TypeError: Failed to fetch is a tricky one  as it doesn't give much info. The error is often due to "CORS" errors (cross-origin). so if the file is on a different server than the host that jbrowse is on, then there are CORS security rules that come into play (and the server that hosts the files must add special CORS headers to allow cross origin access)

i intend to make a better error message than this someday

the easiest thing to do is to host the files and the jbrowse instance on the same server if possible

**Reply by cxbiosyn** [reply]
the file is from the same server as jbrowse actually
the demo page is https://xliugroup.com:39000/, and for admin config, I used the 39001 port, and currently still via IP:39001?adminKey=...

our lab website was made via wordpress and I used a plugin from wordpress to upload the file, the URL looks like this: 
https://xliugroup.com/shared-files/1331/?GCF_000009885.1_ASM988v1_genomic.fa.gz

these are all at the same server, is it (the error) due to the config page not using the domain name(xliugroup.com) but the IP?

Regards
Chenxi

**Reply by cmdcolin** [reply]
when it is on a different port it is subject to the rules of CORS also (see here for same-origin policy, CORS is anything that is not same origin https://developer.mozilla.org/en-US/docs/Web/Security/Same-origin_policy)

 I also do not recommend using the admin-server in production (see warning here https://jbrowse.org/jb2/docs/tutorials/config_gui/#starting-jbrowse-2-admin-server)

instead i would put jbrowse 2 in some static content folder (jbrowse 2 does not have any "server side code", it is just "static html,js,css" so you can put all the jbrowse files in a folder in e.g. /var/www/html/jbrowse to put it in the apache2/nginx main html directory, which just serves static files.
